### ISSUE TITLE
Category Action Plugins

### ISSUE THREAD

**Issue by sivel** [ansible/proposals#55]
# Proposal: Category Action Plugins

*Author*: Matt Martz <@sivel>

*Date*: 2017/03/xx

- Status: New
- Proposal type: core design
- Targeted Release: 2.4
- Associated PR: N/A
- Estimated time to implement: days to weeks

## Motivation

We often have a need to apply very similar action plugins to a category or "class" of modules, most specifically, network modules.

If we had a mechanism to load an action plugin for a category of modules, we could reduce code, and make it easier to impact a whole category of modules.

### Problems
What problems exist that this proposal will solve?
- Reduce duplicated code in action plugins
- Make it easier for a group of modules to share action plugin code, for pulling in runtime info and injecting it into the category of modules

## Solution proposal
This is a still a little fuzzy.  My immediate idea involves a somewhat fundamental change in the way that the plugin loaders work.  I'd really love to start passing around objects, that contain much more data.

But in the end, we would need to have a way, to map a module, to a category, either by utilizing the path based categories we have now, or by embedding categories into the modules. The plugin loaders, potentially those for modules/action plugins, could be merged, with methods or properties describing module location and action plugin location if such a thing exists.

In any case, to simplify that, if we could just have a way to determine what "category" a module was in, we could search for the associated action plugin for that category.  This would be a fallback likely, preferring an exact match first, eventually falling back to "normal".  1) specific, 2) category, 3) normal

## Dependencies (optional)
N/A

## Testing (optional)
Tests would be good

## Documentation (optional)
Documentation would be good

## Anything else?
A real implementation? I'm 100% sure I am not thinking of something.

**Comment by bcoca** [comment]
So I was toying with a possible implementation, based on the current 'name matching' with fallbacks, so each action plugin could declare a list of 'criteria' to be run against the modules:

```self.criteria = ( "name.matches('eol_.*) or path.endswith('cloud/eol')", "other stuff: ) ```

then in executor we can 👍 
```
 if module.name in action_plugins:
    # use action plugin of same name
else:
     for action in action_plugins:
         if action.criteria(module):
               #use matched action plugin
               break
      else:
            # use normal handler
```

**Comment by sivel** [comment]
I think that is roughly equivalent to my idea.  Effectively I was thinking of using the module hierarchy to provide action plugins.

Something like `network/ios/ios_command.py` would look for an `ios_command` action plugin, then `ios` and then `network` and then all the way back to `normal`.

**Comment by rcarrillocruz** [comment]
I like the fallback mechanism, how does it sound to also consume criteria from module metadata ?

i.e. not just hardcoding the path of category selection, but to also allow a module developer to inject its own behaviour.

**Comment by sivel** [comment]
My biggest concern with metadata, is I'm not a huge fan of what I will refer to as introspection, we are having to inspect code to see how to run that code.  It's not exactly efficient, and I know we do things like this already, I just wish we didn't.

**Comment by alikins** [comment]
I'm in favor of sub packages for plugins that have deps that don't fall into the same set of deps as 'ansible'.
ie, ansible-plugins-filters-jmespath as subpackage, with a dep on python-jmespath.

I would make that subpackage include the plugin code, as opposed to just being a deps-only package.
Plugin loader exists so that can be possible, so might as well use it. That also opens the door for letting distros choose to not include/build some subpackage if they can't provide the deps. I am thinking primarily about 'pycrypto' vs 'cryptography' as the example here, though that is currently not isolated to plugins, but
it could be. (See https://github.com/ansible/ansible/pull/25084 for reference)

The main 'ansible' package could 'suggest' ansible-plugins-filters-jmespath, but shouldn't require it. At least if the intention is to provide a path for user to not require 'python-jmespath' for 'ansible' install.

**Comment by bcoca** [comment]
@rcarrillocruz also with module metadata you are inverting the normal execution handling (executor > action > module) to (executor > module > action > module?) ... I really don't think this is something we should allow.

**Comment by rcarrillocruz** [comment]
Agreed, probably default to current action plugin resolving behaviour, and have a toggle in config that allows module-side setting ?

**Comment by abadger** [comment]
I don't think that having a toggle is wise.  then you split the platform.  Here's modules that work with Ansible-toggle-set-to-true  and here's modules that work with Ansible-toggle-set-to-false.  But they don't mix.

I think inverting the handling is the inevitable consequence of wanting this feature.  So if we decide we don't want the feature, we should discard it.  If we do want it, then we have to acknowledge that it is the module calling the shots here, not the action plugin.

Out present hardcoded thing already inverts the relationship -- the module name is a form of metadata that the module is control of, not the action plugin.  This goes slightly further, though -- right now the controller side has a hardcoded mapping of metadata=network:  network_action_plugin.  This proposal would allow that mapping to be drawn by the modules as well as simply the categorization.

We could make something in between, mapping is created and expanded upon controller side by the action plugins.  Then the modules say they need a certain action_feature in order to run.

#### Example Setup
normal action plugin has a field that says:
```
action_handler: "*"
```
bigip action plugin has a field that says:
```
action_handler: "bigip"
```

bigip_template module has in its metadata:
```
{ "metadata_version":  "1.2",
  "action_handler": "bigip"
}

poorly_written plugin has in its metadata:
```
{ "metadata_version": "1.2",
  "action_handler": "nope_does_not_exist"
}

#### Running module with action_handler: handled on the controller side

* Running ```ansible -m bigip_template``` will lookup the metadata for the bigip_template module.
* Ansible will see the action_handler of bigip 
* Ansible will read the metadata for the action plugins
* Ansible will use the metadata to construct a map of action_handler fields to action handlers
* Ansible will load the bigip action plugin because it registered itself as the bigip action_handler.
* Ansible will run the bigip_template module via the bigip action plugin

#### Running module with action_handler: not handled on the controller side

* Running ``` ansible -m poorly_written``` will lookup the metadata for the poorly_written module.
* Ansible will see the action_handler of bigip 
* Ansible will not find an action_handler for nope_does_not_exist
* Ansible will error, unable to run the module

#### Running module with no specified action_handler

* Running ```ansible -m ping``` will lookup the metadata for the ping module
* Ansible will see no action_handler
* Ansible will find the action_handler for "*" is normal
* Ansible will run the ping plugin via the normal action handler.

**Comment by abadger** [comment]
I think I said this in the meeting but recording it here too. I'm against the criteria idea.  It introduces complexity where it isn't needed.

**Comment by bcoca** [comment]
Closing this as collections have basically made the 'category' irrelevant
